support/support.go

```diff
@@ -0,0 +1,227 @@
+package support
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"strings"
+
+	"golang.org/x/xerrors"
+
+	"github.com/google/uuid"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+	"github.com/coder/coder/v2/coderd/rbac"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+// Bundle is a set of information discovered about a deployment.
+// Even though we do attempt to sanitize data, it may still contain
+// sensitive information and should thus be treated as secret.
+type Bundle struct {
+	Deployment Deployment `json:"deployment"`
+	Network    Network    `json:"network"`
+	Workspace  Workspace  `json:"workspace"`
+	Logs       []string   `json:"logs"`
+}
+
+type Deployment struct {
+	BuildInfo    *codersdk.BuildInfoResponse `json:"build"`
+	Config       *codersdk.DeploymentConfig  `json:"config"`
+	Experiments  codersdk.Experiments        `json:"experiments"`
+	HealthReport *codersdk.HealthcheckReport `json:"health_report"`
+}
+
+type Network struct {
+	CoordinatorDebug string                                 `json:"coordinator_debug"`
+	TailnetDebug     string                                 `json:"tailnet_debug"`
+	NetcheckLocal    *codersdk.WorkspaceAgentConnectionInfo `json:"netcheck_local"`
+	NetcheckRemote   *codersdk.WorkspaceAgentConnectionInfo `json:"netcheck_remote"`
+}
+
+type Workspace struct {
+	Workspace        codersdk.Workspace           `json:"workspace"`
+	BuildLogs        []codersdk.ProvisionerJobLog `json:"build_logs"`
+	Agent            codersdk.WorkspaceAgent      `json:"agent"`
+	AgentStartupLogs []codersdk.WorkspaceAgentLog `json:"startup_logs"`
+}
+
+// Deps is a set of dependencies for discovering information
+type Deps struct {
+	// Source from which to obtain information.
+	Client *codersdk.Client
+	// Log is where to log any informational or warning messages.
+	Log slog.Logger
+	// WorkspaceID is the optional workspace against which to run connection tests.
+	WorkspaceID uuid.UUID
+	// AgentID is the optional agent ID against which to run connection tests.
+	// Defaults to the first agent of the workspace, if not specified.
+	AgentID uuid.UUID
+}
+
+func DeploymentInfo(ctx context.Context, client *codersdk.Client, log slog.Logger) Deployment {
+	var d Deployment
+	bi, err := client.BuildInfo(ctx)
+	if err != nil {
+		log.Error(ctx, "fetch build info", slog.Error(err))
```

Comment:
I see the approach you implemented here. It is appropriate, but maybe we could keep a field for bundling errors within the bundle? It would make debugging easier when coderd returns HTTP 500, and the client error is saved.

Reply:
I'll see if there's a nice generic solution to this in a follow-up.

Reply:
Absolutely 👍 This is something that can be done in a follow up.